xdsl/dialects/onnx.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from abc import ABC
+
+from xdsl.dialects.builtin import AnyTensorType, SSAValue
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    OpResult,
+)
+from xdsl.irdl import (
+    IRDLOperation,
+    irdl_op_definition,
+    operand_def,
+    result_def,
+)
+from xdsl.utils.exceptions import VerifyException
+
+
+class ElementwiseBinOpBase(IRDLOperation, ABC):
+    """Base class for element-wise binary operations on tensors with Numpy-style broadcasting."""
+
+    lhs = operand_def()
+    rhs = operand_def()
+    res: OpResult = result_def(AnyTensorType)
+    assembly_format = "`(` $lhs `,` $rhs `)` attr-dict `:` `(` type($lhs) `,` type($rhs) `)` `->` type($res)"
+
+    def __init__(
+        self,
+        lhs: SSAValue,
+        rhs: SSAValue,
+        attributes: dict[str, Attribute] = {},
+    ):
+        super().__init__(
+            operands=[lhs, rhs],
+            attributes=attributes,
+            result_types=[[]],
+        )
+
+    def verify_(self) -> None:
+        # Check that the arguments are broadcastable (using Numpy semantics) and that the result type is correct.
+        res_shape: list[int] = []
+        # Iterate over the shapes in reverse order and compute the result shape.
+        lhs_shape: list[int] = [x.data for x in self.lhs.type.shape]
```

Comment:
Is there a better way to get the shapes of tensor types (I hope so)?

Reply:
Unfortunately not, we have to use isinstance checks for this
```suggestion
            if not isinstance(lhs_type := self.lhs.type, ShapedType):
                raise ValueError(
                    "onnx binary operation operands must be of type ShapedType"
                )
        lhs_shape: list[int] = [x.data for x in lhs_type.shape]
```

Reply:
Actually, you should be able to just write:
`lhs_shape = lhs_type.get_shape()`

It seems that `lhs` and `rhs` are not constrained to be a `Tensor` or a `ShapedType` in their definitions, is that expected?
Actually, are all operands and results supposed to be the same type? Here is an example of how we do it in xDSL: https://github.com/xdslproject/xdsl/blob/2ccde61cb96a8e8ea14c1313ebd6e5df514d84c0/xdsl/dialects/arith.py#L173

Reply:
The `lhs` and `rhs` do not have to be the same type, however, they should have the same element type if they are tensors (e.g., all have to have `int32` elements, or be an `int32` scalar) and the shapes must be broadcastable (what the current validation checks). 

I will add a check for matching element types, which is easy enough; but I don't know how to create them in a way that allows scalar or any shaped type (including tensor).

Reply:
Here is an example on how you could make this work:

```
class BinaryOperation(IRDLOperation):
  T = ConstraintVar("T", Attribute)
  lhs = operand_def(T | VectorOrTensorOf[T])
  rhs = operand_def(T | VectorOrTensorOf[T])
  res = result_def(T | VectorOrTensorOf[T])
```
This will ensure that `lhs`, `rhs`, and `res` are either a `Vector`, a ranked or unranked `Tensor`, and that they all have the same element type.

Making this work for `ShapedType` is a bit more complex, because we are not handling `ShapedType` well currently :/

Reply:
The onnx spec itself actually doesn't allow scalars (I double checked), just ranked tensors, so perhaps something like this is more appropriate I think (similar to `arith`), can you confirm?
```python
_T = TypeVar("_T", bound=Attribute)

class ElementwiseBinOpBase(IRDLOperation):
    """Base class for element-wise binary operations on tensors with Numpy-style broadcasting."""

    T = Annotated[Attribute, ConstraintVar("T"), _T]
    lhs = operand_def(TensorType[T])
    rhs = operand_def(TensorType[T])
    res = result_def(TensorType[T])
  ```
  
In `verify_` the type of `lhs.type` is `xdsl.dialects.builtin.TensorType`, however vscode (and pyright) are not able to statically determine that, and so I still get errors when trying to access and use the shapes. Not sure how to fix those, type hints (e.g., `lhs_shape: tuple[int] = self.lhs.type.get_shape()`) are not enough. So, how can I get the shapes of the operands and result without pyright errors?

Reply:
I'm afraid you still need to add the snippet I suggested above, or something like it

Reply:
I did add the snippet above (and variations) but cannot get types resolved statically, sorry if this is basic, but am kind of out of ideas of what to try to make pyright happy. I can push the latest if it helps.

Reply:
Actually, never mind, I had a silly mistake. I was able to get rid of the pyright errors.

Reply:
Yes, your snippet seems correct!
The main issue here is that you cannot really statically guarantee that `self.lhs` is a `TensorType`, because if you are in the middle of a rewrite, then you cannot guarantee that your operation verifies.
I'm not completely sure how to fix that currently, to be honest, MLIR has now the concept of `TypedValue` that we could use, but I don't have an easy way of integrating that yet without making it a hassle to use :/